docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,215 @@
+data model is actually richer than JSON, with types for byte arrays, various
+precision integers and floats etc.). An alternative approach would be to write
+a custom CRDT for each data type we want to replicate, see [##Alternative
+Approaches] for a discussion of this design.
+
+## Message storage and transmission
+
+The property of CRDTs that we care about is strong eventual consistency: any
+two nodes which have received the same set of messages will merge to the same
+state. A major part of the design of collaborative objects therefore is how we
+store and transmit these messages. Messages in this context are just binary
+blobs which need to be passed to the CRDT implementation in any order.
+
+We add a new entry to the `signed_refs` of a project at
+`refs/namespaces/<project>/remotes/refs/remotes/<peer>/rad/collaborative-objects`.
+This points to a tree with the following structure:
+
+```
+.
+|--<object 1 ID>
+|  |--manifest
+|  |--schema
+|  |  |--<schema change 1>
+|  |  |--<schema change 2>
+|  |--<change 1>
+|  |--<change 2>
+|--<object 2 ID>
+|  ...
+```
+
+We refer to a directory in this tree as an "object directory" and the tuple 
+`(manifest, [schema changes], [changes])` as an "object state".
+
+Objects are created with a unique identifier at creation time. This identifier
+is the name of the directory within the `collaborative_objects` tree where the
+data for this object is kept. Each time we make a change to the object we store
+the new change (which is a binary blob) in the object directory and create a
+new commit. We also store a manifest file which contains the name of the data
+type of this object. This file is never changed and is used to allow
+applications to enumerate collaborative objects of a particular type. Type
+names should be some kind of human readable identifier - similar to XML
+namespaces.
+
+Each object is also created with a schema, this is a JSON schema document which
+is itself encoded as an automerge document within the `schema` directory. This
+allows for changes to the schema (as long as they are backwards and forwards
+compatible) to be distributed as additional changes within the `<object
+ID>/schema` directory.
+
+### Viewing the tracking graph
+
+We need to load the state of objects which multiple people within the
+tracking graph are collaborating on. Given a particular object ID we can load
+all the changes from each remote we have replicated and merge them all to
+obtain the final state of the object. With many remotes this could become
+expensive and there are opportunities to do some indexing ahead of time. A more
+pressing concern is that we want to ensure that the schema is respected, more
+on this later.
+
+
+### Updating objects
+
+To make a change to an object we load the existing messages for an object. The 
+application developer provide us with the binary representation of the change
+to that object. We apply the change and ensure that the new object state still
```

Comment:
Unfortunately not, the change the application developer would provide is (from our perspective) just some bytes. It has to be applied to a document to mean anything. This does open some attack vectors via creating large changes which take a long time to apply.

In this model checking the schema before applying the change would be the responsibility of the CRDT implementation. I'm not aware of any implementation which does this right now, although it is on a list of future work for Automerge so it would be possible to implement it in this manner in the future (and without any protocol changes to boot).

Reply:
Could you maybe give me an example where a change comes in and modifies an object but invalidates the schema? I think that would help me understand this issue a bit better.

Reply:
Imagine we have a schema like the following:

```json
{
    "type": "object",
    "properties": {
        "title": {
            "type": "string",
            "maxLength": 100
        }
    }
}
```

And the current state of the document in our view of the repo is:

```json
{
    "title": "A big imporatnt issue"
}
```

In this case let's say our object directory looks like this:

```
.
|--<schema>
|  |--<schema change>
|--<change 1>
|--<manifest>
```

i.e there is a single binary change file which when loaded resolves to the "current state" I mentioned above.

From `librad`s perspective what we do is this:

- Load the schema changes using automerge to resolve the schema. Effectively we have `schema = Automerge.load(<schema change bytes>)`
- Load the existing state of the object using automerge, `object = Automerge.load(<change 1 bytes>)`
- Pass the automerge document to the application, which has it's way with it and returns a new change, let's call it `<change 2 bytes>`
- The application passes the change back to us, we apply it to the the document so we have `updatedObject = Automerge.apply(object, <change 2 bytes>)`
- Now we check that `updatedObject` is valid with respect to the schema
    - If it is, we compact all the changes and save them in the object directory
    - If it is not then we return an error to the application

So if the application were to return a change which set the title of the object to a string longer than 100 characters then we would return an error.